lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -0,0 +1,216 @@
+prop_fresh Pool{gap,addresses} =
+    takeWhile (== Unused) end == replicate gap Unused
+  where
+    end = map snd $ List.sortOn (Down . fst) $ Map.elems addresses
+
+-- | Internal invariant:
+-- All 'addresses' in the pool have been generated from their index
+-- via the pool 'generator'.
+prop_generator :: Eq addr => Pool addr ix -> Bool
+prop_generator Pool{generator,addresses} =
+    and $ Map.mapWithKey isGenerated addresses
+  where
+    isGenerated addr (ix,_) = generator ix == addr
+
+-- | Internal invariant: The pool satisfies all invariants above.
+prop_consistent :: (Ord ix, Enum ix, Eq addr) => Pool addr ix -> Bool
+prop_consistent p =
+    all ($ p) [prop_sequence, prop_gap, prop_fresh, prop_generator]
+
+{-------------------------------------------------------------------------------
+    Pretty printing
+-------------------------------------------------------------------------------}
+instance Buildable (Pool addr ix) where
+    build pool = "AddressPool "
+        <> "{ " <> build (size pool) <> " addresses"
+        <> ", gap = " <> build (gap pool)
+        <> "}"
+
+instance (Show addr, Show ix) => Show (Pool addr ix) where
+    show pool = "AddressPool"
+        <> "{ generator = <<function>>"
+        <> ", gap = " <> show (gap pool)
+        <> ", addresses = " <> show (addresses pool)
+        <> "}"
+
+{-------------------------------------------------------------------------------
+    Address Pool, operations
+-------------------------------------------------------------------------------}
+-- | Create a new address pool.
+new :: (Ord addr, Enum ix) => (ix -> addr) -> Int -> Pool addr ix
+new generator gap
+    = ensureFresh (toEnum 0) $ Pool{ generator, gap, addresses = Map.empty }
+
+-- | Replace the collection of addresses in a pool,
+-- but only if this collection satisfies the necessary invariants
+-- such as 'prop_sequence' etc.
+load
+    :: (Ord addr,  Ord ix, Enum ix)
+    => Pool addr ix -> Map addr (ix,AddressState) -> Maybe (Pool addr ix)
+load addrs pool0 = if prop_consistent pool then Just pool else Nothing
+  where pool = loadUnsafe addrs pool0
+
+-- | Replace the collection of addresses in a pool,
+-- but skips checking the invariants.
+loadUnsafe :: Pool addr ix -> Map addr (ix,AddressState) -> Pool addr ix
+loadUnsafe pool addrs = pool{ addresses = addrs }
+
+-- | Remove all previously discovered addresses,
+-- i.e. create a new pool with the same 'generator' and 'gap' as the old pool.
+clear :: (Ord addr, Enum ix) => Pool addr ix -> Pool addr ix
+clear Pool{generator,gap} = new generator gap
+
+-- | Look up an address in the pool.
+lookup :: Ord addr => addr -> Pool addr ix -> Maybe ix
+lookup addr Pool{addresses} = fst <$> Map.lookup addr addresses
+
+-- | Number of addresses cached in the pool.
+size :: Pool addr ix -> Int
+size = Map.size . addresses
+
+-- | Given an index, retrieve the next index that is still in the pool.
+next :: Enum ix => Pool addr ix -> ix -> Maybe ix
```

Comment:
also, what about adding next ix of Used address, Unused address, or does not matter what address?

Reply:
in case of next ix of Used...I would also think about having possibility to get ix of first (the smallest possible ix overall), or the first next AFTER the last was used. those two situations are possible - example ->  https://github.com/input-output-hk/cardano-wallet/pull/3068#discussion_r774528985

Reply:
Actually, I was not quite sure whether to include this function at all. 🤔 My original intention was to use it for "light-wallet-style" address discovery, where each address is queried in turn, and discovery stops when the `next` function returns `Nothing`.

Extending this function to distinguish `Used` or `Unused` status may be useful for generating change addresses. Currently, change addresses use `size` and `gap` directly, but I do like the idea of moving all "knowledge about the pool gap" into this module. I would like to keep `next` as it is now, and revisit this issue in a future PR.

Reply:
What I am saying is that:
- we can use meticulously next here, and we will have state like that:
```
Used Used Used Unused (gap times)
```
Here, `next` will give `Just 3`
But, we can have people using hardware wallet and then they could restore in cardano-wallet basedon mnemonic. And they can MANUALLY have the following state:
```
Used Unused Used Used Unused (gap times)
```
So I just want to know and understand what we are going to get when `next` is called: (a) Just 1 OR (b) Just 4 ?
So the retrieve the Unused ix with the lowest ix, OR retrieve the first Unused after Used with the highest ix. 
You see my point?

Reply:
Ah. The intention is that `next ix` does not care about the 'Used' or 'Unused' status at all, and will simply return `succ ix` most of the time.

This is useful for address discovery in a light client setting, where the discovery process goes like this: Start with `ix = 0`, query the explorer for the corresponding address, continue with `next ix` until it returns `Nothing`.

The `next` function is not intended to be used to generate change addresses where the `Used` and `Unused` status matters.

I will update the documentation accordingly and maybe use a different name. 🤔